README
NAME
    Devel::SearchINC - Loading Perl modules from their development
    directories

VERSION
    version 2.101440

SYNOPSIS
      # export PERL5OPT=-MDevel::SearchINC=/my/dev/dir
      # now use your new modules

DESCRIPTION
    When developing a new module, I always start with a standard skeleton
    distribution directory. The directory structure is such, however, that
    you have to install the module first (with "make install") before you
    can use it in another program or module. For example, bringing in a
    module like so:

        use My::Module;

    requires the module to be somewhere in a path listed in @INC, and the
    relative path is expected to be "My/Module.pm". However, "h2xs" creates
    a structure where the module ends up in "My/Module/Module.pm".

    This module tries to compensate for that. The idea is that you "use()"
    it right at the beginning of your program so it can modify @INC to look
    for modules in relative paths of the special structure mentioned above,
    starting with directories specified along with the "use()" statement
    (i.e. the arguments passed to this module's "import()").

    This is useful because with this module you can test your programs using
    your newly developed modules without having to install them just so you
    can use them. This is especially advantageous when you consider working
    on many new modules at the same time.

    To automatically make your development modules available to all your
    scripts, you can place the following in your ".bashrc" (or your shell
    initialization file of choice):

      export PERL5OPT=-MDevel::SearchINC=/my/dev/dir

    Tilde expansion is also performed.

    When this module is first run, that is, when perl first consults @INC,
    all candidate files are remembered in a cache (see "build_cache()"). A
    candidate file is one whose name ends in ".pm", is not within a
    directory called "t", "CVS", ".svn", ".git", "skel" or "_build", and is
    within a directory called "lib", "blib/lib" or "blib/arch". This is a
    long-winded way of saying that it tries to find your perl module files
    within standard development directories.

    If a directory contains a file named "INC.SKIP", this directory will be
    skipped.

    Note that there is a small limitation for the "PERL5OPT" approach:
    development modules can't be loaded via "-M" on the perl command line.
    So the following won't work:

      $ export PERL5OPT=-MDevel::SearchINC=/my/dev/dir
      $ perl -MMy::Brand::New::Module -e'print "hello world\n"'

    This is because "PERL5OPT" options are processed after the actual
    command line options.

    Also, the "PERL5OPT" variable is ignored when Taint checks are enabled.

FUNCTIONS
  build_cache
    Called during "import()", this subroutine builds a cache of the modules
    it finds. This way every time a module is "use()"d, we can just look at
    the cache. This does mean that if you add, change or delete a module
    during the run-time of a script that uses "Devel::SearchINC", that
    script won't notice these changes. If necessary, you could re-run
    "build_cache()".

MULTIPLE DEVELOPMENT DIRECTORIES
    You can have multiple development directories. Just list them when using
    this module:

      use Devel::SearchINC qw(/my/first/dir my/second/dir);

    or

      perl -MDevel::SearchINC=/my/first/dir,/my/second/dir

    You can also use semicolons or colons instead of commas as delimiters
    for directories.

    "perlrun" details the syntax for specifying multiple arguments for
    modules brought in with the "-M" switch.

SPECIAL OPTIONS
    In addition to development directories, there are special options that
    can be passed to this module:

    "-debug"
        This turns on debugging; see below.

    "-clear"
        This option clears all paths that were previously set. This is
        useful when you have a "PERL5OPT" environment variable but want to
        use this module from within your code as well and want to override
        the environment setting. For example:

            export PERL5OPT=-MDevel::SearchINC=/path/to/dev

        then:

            #!/usr/bin/env perl

            use warnings;
            use strict;
            use Devel::SearchINC qw(-clear /path/to/other/dir);
            ...

DEBUGGING THIS MODULE
    By using "-debug" as one of the development directories, you can turn on
    debugging. Note that despite the leading colon, this has nothing to do
    with "Exporter" semantics. With debugging activated, this module will
    print detailed information while trying to find the requested file.

    For example

      use Devel::SearchINC qw(/my/first/dir my/second/dir -debug);

    or

      perl -MDevel::SearchINC=/my/first/dir,-debug,/my/second/dir

    The "-debug" option can be specified anywhere in the list of development
    directories.

INSTALLATION
    See perlmodinstall for information and options on installing Perl
    modules.

BUGS AND LIMITATIONS
    No bugs have been reported.

    Please report any bugs or feature requests through the web interface at
    <http://rt.cpan.org/Public/Dist/Display.html?Name=Devel-SearchINC>.

AVAILABILITY
    The latest version of this module is available from the Comprehensive
    Perl Archive Network (CPAN). Visit <http://www.perl.com/CPAN/> to find a
    CPAN site near you, or see
    <http://search.cpan.org/dist/Devel-SearchINC/>.

    The development version lives at
    <http://github.com/hanekomu/Devel-SearchINC/>. Instead of sending
    patches, please fork this project using the standard git and github
    infrastructure.

AUTHOR
      Marcel Gruenauer <marcel@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2004 by Marcel Gruenauer.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Perl extension Devel-SearchINC

2.101440  2010-05-24 20:50:20 Europe/Vienna
    - fixed Makefile.PL's BUILD_REQUIRES handling by using latest Dist::Zilla
    - more realistic SYNOPSIS

2.101410  2010-05-21 12:40:11 Europe/Vienna
    - set major version to 2 so version continuity on CPAN is preserved.

1.101400  2010-05-20 14:04:46 Europe/Vienna
    - removed bin/distfind; now lives in its own App-distfind distribution

1.101040  2010-04-14 18:10:19 Europe/Vienna
    - the :debug option is now -debug
    - added -clear option
    - development directories and options can be separated by colons as well
    - improved tests

1.101000  2010-04-10 18:35:28 Europe/Vienna
    - added bin/distfind, adapted from the now-deprecated Dist::Joseki
    - removed unused variables

1.100850  2010-03-26 21:34:37 Europe/Vienna
    - converted the distribution to Dist::Zilla-style

1.39 Sat 2010.02.06 14:46:06 CET (Marcel Gruenauer <marcel@cpan.org>)
    - set broad untaint_pattern for File::Find, as that gets it wrong on OS X
      (its pattern doesn't allow spaces in filenames)

1.38 Tue Aug  4 19:09:11 WEST 2009 (Marcel Gruenauer <marcel@cpan.org>)
    - more output under :debug
    - prune a directory if it contains INC.SKIP

1.37  Thu Nov  6 23:09:35 CET 2008 (Marcel Gruenauer <marcel@cpan.org>)
    - caused a bug where a program running under -T caused an "Insecure
      dependency in chdir while running with -T switch" error

1.36  Tue Nov  4 12:49:31 CET 2008 (Marcel Gruenauer <marcel@cpan.org>)
    - build_cache() just returns if no paths were set to avoid a warning from
      File::Find::find(); had CPAN tester failures because of that; my own
      tests worked because I used Devel::SearchINC in $PERL5OPT...

1.35  Thu Oct 30 16:32:55 CET 2008 (Marcel Gruenauer <marcel@cpan.org>)
    - If a module is found, set its entry in %INC to avoid the
      '/loader/0x1234567/Foo/Bar.pm' effect, so it works nicely with
      Devel::NYTProf

1.34  Thu Jul 25 15:27:43 CEST 2008 (Marcel Gruenauer <marcel@cpan.org>)
    - Changed dist style and Changes back to standard. People didn't like it -
      the nail that sticks out gets hammered down.
    - Added standard test files; this will also help with CPANTS.
    - Devel::SearchINC now uses File::Find::find() like
      Devel::SearchINC::FindLib does, and caches the result during the first
      pass. So this is more flexible than the previous version and allows you
      to use Devel::SearchINC with complex development directory setups,
      without the disadvantage of Devel::SearchINC::FindLib cluttering up your
      @INC.

1.33  Thu, 29 May 2008 12:04:28 -0000 (Marcel Gruenauer <marcel@cpan.org>)
    - Converted Changes file to YAML style
    - .shipit: added Twitter step
    - Makefile.PL: added auto_install() and process_templates()
    - lib/*: converted to template
    - updated MANIFEST
    - Devel::SearchINC::FindLib now also skips skel/
    - added t/perlcriticrc
    - updated MANIFEST
    - moved generated documentation of interited methods to after documentation
      of this class' methods so the manpage is more readable
    - Devel::SearchINC::FindLib can now handle semicolon-separated paths as
      well
    - .shipit: new svk.tagpattern
    - Devel::SearchINC::FindLib: also deal with blib; ignore _build
    - .shipit: fixed svk.tagpattern
    - tags: NEWFEATURE

1.32  Thu, 18 Oct 2007 10:09:39 +0200 (Marcel Gruenauer <marcel@cpan.org>)
    - fixed version requirement of Test::More

1.31  Wed, 03 Oct 2007 17:27:25 +0200 (Marcel Gruenauer <marcel@cpan.org>)
    - changed the $VERSION of all modules to '1.31'.
    - switched back to Makefile.PL and to Module::AutoInstall
    - uses Module::Install::StandardTests

1.03  Thu, 17 Feb 2005 23:53:39 +0100 (Marcel Gruenauer <marcel@cpan.org>)
    - added Devel::SearchINC::FindLib
    - switched to Build.PL

1.02  Sun, 01 Sep 2002 12:06:22 +0200 (Marcel Gruenauer <marcel@cpan.org>)
    - tests now use Test::More and have been reorganized
    - added a debugging option and verbose debugging output
    - previously it tried to open last guess regardless of whether the search
      loop was ended because there were no more paths to search or because it
      found a file. So I introduced a $found variable that takes care of that.

0.03  Thu, 11 Oct 2001 09:46:10 +0200 (Marcel Gruenauer <marcel@cpan.org>)
    - added PERL5OPT documentation (thanks Michael Graham)

0.02  Sat, 19 May 2001 13:39:31 +0100 (Marcel Gruenauer <marcel@cpan.org>)
    - can now also find modules in lib/ subdirectories

0.01  Thu, 15 Mar 2001 14:38:37 +0100 (Marcel Gruenauer <marcel@cpan.org>)
    - original version


t/lib/C/D/lib/E.pm
package E;
our $VERSION = '1.100850';
sub answer { 42 }
1;


t/lib/C/D/lib/C/D/F.pm
package C::D::F;
our $VERSION = '1.100850';
sub answer { 42 }
1;


lib/Devel/SearchINC.pm
use 5.008;
use strict;
use warnings;

package Devel::SearchINC;
BEGIN {
  $Devel::SearchINC::VERSION = '2.101440';
}
# ABSTRACT: Loading Perl modules from their development directories
use Data::Dumper;
use File::Find;

sub build_cache {
    our %cache = ();
    our @PATHS;
    return unless @PATHS;
    our $DEBUG;

    # Programs run with -T cause a "Insecure dependency in chdir while running
    # with -T switch" warning, so untaint directory names.
    find(
        {   untaint         => 1,
            untaint_pattern => qr|^(.+)$|,  # File::Find gets this wrong on OS X
            untaint_skip    => 1,
            wanted          => sub {
                warn "dir [$File::Find::name]\n" if $DEBUG && -d;
                if (-d && /^(t|CVS|\.svn|\.git|skel|_build)$/) {
                    warn "$File::Find::name dir will be pruned\n" if $DEBUG;
                    $File::Find::prune = 1;
                    return;
                }
                if (-d && -e "$File::Find::name/INC.SKIP") {
                    warn "$File::Find::name dir contains INC.SKIP; pruned\n"
                      if $DEBUG;
                    $File::Find::prune = 1;
                    return;
                }
                if (-d && $_ eq 'lib') {
                    push our @inc => $File::Find::name;
                    return;
                }
                return unless -f && /\.pm$/;
                if ($File::Find::name =~ m!.*/(?:lib|blib/(?:lib|arch))/(.*)!) {
                    $cache{$1} ||= $File::Find::name;
                }
              }
        },
        @PATHS
    );
    warn "cache:\n", Dumper \%cache if $DEBUG;
}

BEGIN {
    unshift @INC, sub {
        my ($self, $file) = @_;
        our %cache;
        our $DEBUG;
        unless (exists $cache{$file}) {
            printf "%s: cache miss <%s>\n", __PACKAGE__, $file
              if $DEBUG;
            return;
        }
        printf "%s: found <%s>\n", __PACKAGE__, $cache{$file} if $DEBUG;
        if (open(my $fh, '<', $cache{$file})) {
            $INC{$file} = $cache{$file};
            return $fh;
        }
        printf "%s: can't open <%s>, declining\n", __PACKAGE__, $cache{$file}
          if $DEBUG;
        return;
      }
}

sub import {
    shift;    # we don't need the package name
    our $DEBUG = 0;
    my @p =
      map { s/^~/$ENV{HOME}/; $_ }
      map { split /\s*[,;:]\s*/ } @_;
    our @PATHS;
    for my $path (@p) {
        if ($path eq '-debug') {
            $DEBUG = 1;
            next;
        }
        if ($path eq '-clear') {
            @PATHS = ();
            next;
        }
        push @PATHS => $path;
    }

    # Build the module cache anew after each import; this is so that if you
    # use PERL5OPT=-MDevel::SearchINC=... and then a program loads it
    # separately with "use Devel::SearchINC '...'" paths from both occasions
    # get respected.
    build_cache();
    warn "paths:\n", Dumper \@PATHS if $DEBUG;
}
1;


__END__
=pod

=head1 NAME

Devel::SearchINC - Loading Perl modules from their development directories

=head1 VERSION

version 2.101440

=head1 SYNOPSIS

  # export PERL5OPT=-MDevel::SearchINC=/my/dev/dir
  # now use your new modules

=head1 DESCRIPTION

When developing a new module, I always start with a standard skeleton
distribution directory. The directory structure is such, however, that you
have to install the module first (with C<make install>) before you can use it
in another program or module. For example, bringing in a module like so:

    use My::Module;

requires the module to be somewhere in a path listed in C<@INC>, and
the relative path is expected to be C<My/Module.pm>. However, C<h2xs>
creates a structure where the module ends up in C<My/Module/Module.pm>.

This module tries to compensate for that. The idea is that you C<use()>
it right at the beginning of your program so it can modify C<@INC> to look
for modules in relative paths of the special structure mentioned above,
starting with directories specified along with the C<use()> statement
(i.e. the arguments passed to this module's C<import()>).

This is useful because with this module you can test your programs using
your newly developed modules without having to install them just so you
can use them. This is especially advantageous when you consider working
on many new modules at the same time.

To automatically make your development modules available to all
your scripts, you can place the following in your C<.bashrc> (or
your shell initialization file of choice):

  export PERL5OPT=-MDevel::SearchINC=/my/dev/dir

Tilde expansion is also performed.

When this module is first run, that is, when perl first consults C<@INC>, all
candidate files are remembered in a cache (see C<build_cache()>). A candidate
file is one whose name ends in C<.pm>, is not within a directory called C<t>,
C<CVS>, C<.svn>, C<.git>, C<skel> or C<_build>, and is within a directory
called C<lib>, C<blib/lib> or C<blib/arch>. This is a long-winded way of
saying that it tries to find your perl module files within standard
development directories.

If a directory contains a file named C<INC.SKIP>, this directory will be
skipped.

Note that there is a small limitation for the C<PERL5OPT> approach:
development modules can't be loaded via C<-M> on the perl command
line.  So the following won't work:

  $ export PERL5OPT=-MDevel::SearchINC=/my/dev/dir
  $ perl -MMy::Brand::New::Module -e'print "hello world\n"'

This is because C<PERL5OPT> options are processed after the actual command
line options.

Also, the C<PERL5OPT> variable is ignored when Taint checks are
enabled.

=head1 FUNCTIONS

=head2 build_cache

Called during C<import()>, this subroutine builds a cache of the modules it
finds. This way every time a module is C<use()>d, we can just look at the
cache. This does mean that if you add, change or delete a module during the
run-time of a script that uses C<Devel::SearchINC>, that script won't notice
these changes. If necessary, you could re-run C<build_cache()>.

=head1 MULTIPLE DEVELOPMENT DIRECTORIES

You can have multiple development directories. Just list them when using
this module:

  use Devel::SearchINC qw(/my/first/dir my/second/dir);

or

  perl -MDevel::SearchINC=/my/first/dir,/my/second/dir

You can also use semicolons or colons instead of commas as delimiters for
directories.

C<perlrun> details the syntax for specifying multiple arguments for
modules brought in with the C<-M> switch.

=head1 SPECIAL OPTIONS

In addition to development directories, there are special options that can be
passed to this module:

=over 4

=item C<-debug>

This turns on debugging; see below.

=item C<-clear>

This option clears all paths that were previously set. This is useful when you
have a C<PERL5OPT> environment variable but want to use this module from
within your code as well and want to override the environment setting. For
example:

    export PERL5OPT=-MDevel::SearchINC=/path/to/dev

then:

    #!/usr/bin/env perl

    use warnings;
    use strict;
    use Devel::SearchINC qw(-clear /path/to/other/dir);
    ...

=back

=head1 DEBUGGING THIS MODULE

By using C<-debug> as one of the development directories, you can turn
on debugging. Note that despite the leading colon, this has nothing to
do with C<Exporter> semantics. With debugging activated, this module
will print detailed information while trying to find the requested file.

For example

  use Devel::SearchINC qw(/my/first/dir my/second/dir -debug);

or

  perl -MDevel::SearchINC=/my/first/dir,-debug,/my/second/dir

The C<-debug> option can be specified anywhere in the list of development
directories.

=head1 INSTALLATION

See perlmodinstall for information and options on installing Perl modules.

=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests through the web interface at
L<http://rt.cpan.org/Public/Dist/Display.html?Name=Devel-SearchINC>.

=head1 AVAILABILITY

The latest version of this module is available from the Comprehensive Perl
Archive Network (CPAN). Visit L<http://www.perl.com/CPAN/> to find a CPAN
site near you, or see
L<http://search.cpan.org/dist/Devel-SearchINC/>.

The development version lives at
L<http://github.com/hanekomu/Devel-SearchINC/>.
Instead of sending patches, please fork this project using the standard git
and github infrastructure.

=head1 AUTHOR

  Marcel Gruenauer <marcel@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2004 by Marcel Gruenauer.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/Devel/SearchINC/FindLib.pm
package Devel::SearchINC::FindLib;

use strict;
use warnings;
use File::Find;


our $VERSION = '1.33';


sub import {
    my $pkg = shift;

    my @found;
    for my $path (@_) {

        # Allow paths to be separated by semicolons as well as colons;
        # splitting along colons is done by perl already.

        my @path =
            map { s/^~/$ENV{HOME}/; $_ }
            split /\s*;\s*/ =>
            $path;

        find(sub {
            return unless -d;

            if (/^t|CVS|\.svn|skel|_build$/) {
                $File::Find::prune = 1;
                return;
            }

            # lib before blib because we're unshifting and if both exist, we
            # want blib to be found first
            if ($File::Find::name =~ m!/(lib|blib/(lib|arch))$!) {
                unshift @found, $File::Find::name;
                $File::Find::prune = 1;
                return;
            }

        }, @path);

        # a simple alphabetical sort is enough for now because 'blib' sorts
        # before 'lib'. If requirements get more complicated, use a
        # Schwartzian Transform

        unshift @INC, sort @found;
    }
    unshift @INC, split /:/ => $ENV{SEARCHINC_FIRST}
        if defined $ENV{SEARCHINC_FIRST};
}


1;


__END__



=head1 NAME

Devel::SearchINC::FindLib - Find lib/ directories and prepend them to @INC

=head1 SYNOPSIS

    Devel::SearchINC::FindLib->new;

=head1 DESCRIPTION

None yet.

=head1 METHODS

=over 4



=back

Devel::SearchINC::FindLib inherits from .

=head1 TAGS

If you talk about this module in blogs, on del.icio.us or anywhere else,
please use the C<develsearchinc> tag.

=head1 VERSION 
                   
This document describes version 1.33 of L<Devel::SearchINC::FindLib>.

=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests to
C<<bug-devel-searchinc@rt.cpan.org>>, or through the web interface at
L<http://rt.cpan.org>.

=head1 INSTALLATION

See perlmodinstall for information and options on installing Perl modules.

=head1 AVAILABILITY

The latest version of this module is available from the Comprehensive Perl
Archive Network (CPAN). Visit <http://www.perl.com/CPAN/> to find a CPAN
site near you. Or see <http://www.perl.com/CPAN/authors/id/M/MA/MARCEL/>.

=head1 AUTHORS

Marcel GrE<uuml>nauer, C<< <marcel@cpan.org> >>

=head1 COPYRIGHT AND LICENSE

Copyright 2004-2008 by the authors.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


=cut



Makefile.PL

use strict;
use warnings;

 BEGIN { require 5.008; } 

use ExtUtils::MakeMaker 6.31;



my %WriteMakefileArgs = (
  'ABSTRACT' => 'Loading Perl modules from their development directories',
  'AUTHOR' => 'Marcel Gruenauer <marcel@cpan.org>',
  'BUILD_REQUIRES' => {
    'Capture::Tiny' => '0',
    'Carp' => '0',
    'English' => '0',
    'File::Temp' => '0',
    'Scalar::Util' => '0',
    'Test::More' => '0.94'
  },
  'CONFIGURE_REQUIRES' => {
    'ExtUtils::MakeMaker' => '6.31'
  },
  'DISTNAME' => 'Devel-SearchINC',
  'EXE_FILES' => [],
  'LICENSE' => 'perl',
  'NAME' => 'Devel::SearchINC',
  'PREREQ_PM' => {
    'Data::Dumper' => '0',
    'File::Find' => '0'
  },
  'VERSION' => '2.101440',
  'test' => {
    'TESTS' => 't/*.t'
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM}; 
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod}; 
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



